src/ott/solvers/quadratic/histogram_transport.py

```diff
@@ -0,0 +1,259 @@
+        errors=errors,
+    )
+
+
+# @jax.tree_util.register_pytree_node_class
+class HistogramTransport:
+  """Histogram Transport solver.
+
+  WARNING: As implemented, this solver assumes uniform marginals,
+  non-uniform marginal solver coming soon!
+
+  Computes the First Lower Bound distance between two distributions.
+  The current implementation requires uniform marginals.
+  If there are an uneven number of points in the two distributions,
+  then we perform a stratified subsample of the distribution of
+  distances to be able to approximate the Wasserstein distance between
+  the local distributions of distnaces.
+
+  Args:
+  `epsilon_1d`: regularization for soft sort. Set to `0.0` for normal sorting
+  `p`: exponent for computing the transport distance betweeen histograms
+  `epsilon`: regularization parameter for the resulting Sinkhorn problem
+  `min_iterations`: minimum iterations for computing Sinkhorn distance
+  `max_iterations`: maximum iterations for computing Sinkhorn distance
+  `cost_fn`: cost function for transport on the real line. If this is
+  provided, the parameter `p` is ignored
+  """
+
+  def __init__(
+      self,
+      epsilon_1d: float = 0.0,
+      p: float = 1.0,
+      epsilon: float = 1.0,
+      min_iterations: int = 10,
+      max_iterations: int = 100,
+      cost_fn: Union[None, TICost] = None,
+  ):
+    self.epsilon_1d = epsilon_1d
+    self.p = p
+    self.epsilon = epsilon
+    self.linear_ot_solver = sinkhorn.Sinkhorn(
+        max_iterations=max_iterations, min_iterations=min_iterations
+    )
+    self.cost_fn = cost_fn
+
+  def __call__(
+      self,
+      prob: quadratic_problem.QuadraticProblem,
+      rng: Optional[jax.random.PRNGKeyArray] = None,
+  ) -> HTOutput:
+    """Run the Histogram Transport solver.
+
+    Args:
+    prob: Quadratic OT problem.
+    rng: Random number key.
+
+    Returns:
+    The Histogram Transport output.
+    """
+    rng = utils.default_prng_key(rng)
+
+    dists_xx = prob.geom_xx.cost_matrix
+    dists_yy = prob.geom_yy.cost_matrix
+    m, n = dists_xx.shape[0], dists_yy.shape[0]
+    min_num_pts = min([m, n])
+    small_indices = jnp.arange(min_num_pts)
+    x_indices = jnp.round(small_indices * (m / min_num_pts)).astype(int)
+    y_indices = jnp.round(small_indices * (n / min_num_pts)).astype(int)
+
+    if self.epsilon_1d <= 0.0:
+      sorted_dists_xx = jax.lax.sort(dists_xx, dimension=-1)
+      sorted_dists_yy = jax.lax.sort(dists_yy, dimension=-1)
+    else:
+      sorted_dists_xx = soft_sort.sort(
+          dists_xx, axis=-1, epsilon=self.epsilon_1d
+      )
+      sorted_dists_yy = soft_sort.sort(
+          dists_yy, axis=-1, epsilon=self.epsilon_1d
+      )
+
+    # Uniformly subsample distances
+    sorted_dists_xx = jnp.take_along_axis(
+        sorted_dists_xx, x_indices[None, :], -1
+    )
+    sorted_dists_yy = jnp.take_along_axis(
+        sorted_dists_yy, y_indices[None, :], -1
+    )
+
+    if self.cost_fn is None:
+      match self.p:
```

Comment:
@Daniel-Packer Python doesn't have `match/case`. Also to simplify, I think the user should supply only the `cost_fn` and not `p` (we have implemented `SqPNorm` and `PNormP`, so it's a bit redundant.

Will give you a full review once the all the code can be parsed.

Reply:
Oh yup. I think Python 3.10 has `match/case`, but that's pretty recent. I'll drop the `p`, good point!